Library/Homebrew/dependency.rb

```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
       all.map(&:name).uniq.map do |name|
         deps = grouped.fetch(name)
         dep  = deps.first
-        tags = deps.flat_map(&:tags).uniq
+        tags = merge_tags(deps)
         option_names = deps.flat_map(&:option_names).uniq
         dep.class.new(name, tags, dep.env_proc, option_names)
       end
     end
+
+    private
+
+    def merge_tags(deps)
+      options = deps.flat_map(&:option_tags).uniq
+      merge_necessity(deps) + merge_temporality(deps) + options
+    end
+
+    def merge_necessity(deps)
+      if deps.any? { |dep| !dep.recommended? && !dep.optional? }
+        # FIXME: Should be `required?`, but `required?` implies `!build?`.
+        [] # Implies required dependency.
+      elsif deps.any?(&:recommended?)
+        [:recommended]
+      else # deps.all?(&:optional?)
+        [:optional]
+      end
```

Comment:
I personally think that my implementation is more readable, because it is expressed on a higher level. I find it easier to reason about all dependencies at once instead of understanding the various state changes that happen in the “loop” you suggest. And I think correctness is (in this case) more important than performance.

Here's my implementation translated into English which hopefully illustrates my point:

- If **any** of the dependencies is a **required** dependency:
  - The merged dependency shall be **required**.
- Otherwise, if **any** of the dependencies is a **recommended** dependency:
  - The merged dependency shall be **recommended**.
- Otherwise, **all** dependencies must be **optional** and:
  - The merged dependency shall be **optional**, too.